### PR TITLE
Fix filename decoding issue for zip files archived by macOS

### PR DESCRIPTION
**Pull request by Wh1isper**
this is how I fix https://github.com/jupyterlab-contrib/jupyter-archive/issues/74

### CODE REVIEW

File: jupyter_archive/handlers.py [hunk 1]
```diff
@@ -223,17 +223,42 @@ async def get(self, archive_path, include_body=False):
...
+                return None
+
+        def unzip(reader, destination):
+            # most of Windows implementations use DOS (OEM) encoding
+            # Mac OS zip utility uses utf-8, but it doesn't set utf-8 bit flags
+            # *nix zip utilities silently uses system encoding(utf-8 generally)
+            with reader as f:
+                for fn in f.namelist():
+                    extreact_path = pathlib.Path(f.extract(fn, path=destination))
+                    correct_filename = try_macos_decode(fn) or try_windows_chinese_decode(fn) or fn
```
Comment (fcollonval) [inline jupyter_archive/handlers.py:245]: Could you change this a bit to use a single function that has the following structure:

```py
def decode_filename(fn):
    try:
        encoded_filename = fn.encode('cp437')
        try:
            # MacOS encoding
            return encoded_filename.decode('utf-8')
        except UnicodeError:
            # Windows encoding
            return encoded_filename.decode('gbk')
    except UnicodeError:
        return fn
```

File: jupyter_archive/handlers.py [hunk 2]
```diff
@@ -223,17 +223,42 @@ async def get(self, archive_path, include_body=False):
...
+                    extreact_path.rename(os.path.join(destination, correct_filename))
```
Comment (fcollonval) [inline jupyter_archive/handlers.py:246]: `destination` is a `Path`, so you can write:

```suggestion
                    extracted_path.rename(destination / correct_filename)
```

File: jupyter_archive/handlers.py [hunk 3]
```diff
@@ -223,17 +223,42 @@ async def get(self, archive_path, include_body=False):
...
+            except UnicodeError:
+                return None
+
+        def unzip(reader, destination):
+            # most of Windows implementations use DOS (OEM) encoding
+            # Mac OS zip utility uses utf-8, but it doesn't set utf-8 bit flags
+            # *nix zip utilities silently uses system encoding(utf-8 generally)
+            with reader as f:
+                for fn in f.namelist():
+                    extreact_path = pathlib.Path(f.extract(fn, path=destination))
```
Comment (fcollonval) [inline jupyter_archive/handlers.py:244]: Fix wording

```suggestion
                    extracted_path = pathlib.Path(f.extract(fn, path=destination))
```

File: jupyter_archive/handlers.py [hunk 4]
```diff
@@ -223,17 +223,42 @@ async def get(self, archive_path, include_body=False):
...
+        if isinstance(archive_reader, zipfile.ZipFile):
+            unzip(archive_reader, archive_destination)
+        else:
+            with archive_reader as archive:
+                archive.extractall(archive_destination)
 
         self.log.info("Finished extracting {} to {}.".format(archive_path, archive_destination))
 
 
+
```
Comment (fcollonval) [inline jupyter_archive/handlers.py:261]:
```suggestion
```

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks @Wh1isper 

I have a couple of suggestions. And could you update the workflow `.github/workflows/build.yml` on all OSes, so the tests will cover this use case.

**Comment by Wh1isper** [comment]
well, I found this is not a good way to solve this problem
but anyone need this temporarily solution can refer my fork

**Pull request closed without merging**